### PR TITLE
Make sure to allocate enough sized waitEventSet

### PR DESCRIPTION
**Pull request by onderkalaci**
With multi-row INSERTs, we might have more sessions than
task*workerCount after few calls of RunDistributedExecution()
because the previous sessions would also be alive.

Instead, re-allocate events when the connectino set changes.

DESCRIPTION: PR description that will go into the change log, up to 78 characters

### CODE REVIEW

File: src/backend/distributed/executor/executor.c [hunk 1]
```diff
@@ -1350,18 +1341,28 @@ RunDistributedExecution(DistributedExecution *execution)
...
 				FreeWaitEventSet(execution->waitEventSet);
+
 				execution->waitEventSet = BuildWaitEventSet(execution->sessionList);
+
+				if (execution->connectionSetChanged)
+				{
+					/* recalculate since the sessions have changed */
+					eventSetSize = EventSetSizeForSessionList(execution->sessionList);
+
+					events = palloc0(eventSetSize * sizeof(WaitEvent));
```
Comment (marcocitus) [inline src/backend/distributed/executor/executor.c]: I think we should do a `pfree` of events here.

File: src/backend/distributed/executor/executor.c [hunk 2]
```diff
@@ -1418,7 +1419,10 @@ RunDistributedExecution(DistributedExecution *execution)
 		 */
 		UnclaimAllSessionConnections(execution->sessionList);
 
-		pfree(events);
+		/*
+		 * We don't need to (and probably shouldn't) free events because
+		 * it might have never been allocated.
+		 */
```
Comment (marcocitus) [inline src/backend/distributed/executor/executor.c]: I don't think we need to explain that we don't do this, because I don't think the reader is necessarily expecting it. Freeing things is relatively unusual in postgres.

File: src/backend/distributed/executor/executor.c [hunk 3]
```diff
@@ -1310,33 +1310,24 @@ void
...
-	/* additional 2 is for postmaster and latch */
-	maxWaitEventCount = execution->totalTaskCount * workerCount + 2;
-
-	/* allocate events for the maximum number of connections to avoid realloc */
-	events = palloc0(maxWaitEventCount * sizeof(WaitEvent));
-
 	PG_TRY();
 	{
 		bool cancellationReceived = false;
+		int eventSetSize = EventSetSizeForSessionList(execution->sessionList);
```
Comment (marcocitus) [inline src/backend/distributed/executor/executor.c]: I'd prefer this to be more explicit to show what we're doing and why:
```
/* additional 2 is for postmaster and latch */
int eventSetSize = list_length(execution->sessionList) + 2;
```

File: src/backend/distributed/executor/executor.c [hunk 4]
```diff
@@ -2465,7 +2469,7 @@ WorkerPoolFailed(WorkerPool *workerPool)
 	ListCell *sessionCell = NULL;
 
 	/* a pool cannot fail multiple times */
-	Assert (!workerPool->failed);
+	Assert(!workerPool->failed);
```
Comment (marcocitus) [inline src/backend/distributed/executor/executor.c]: should maybe be in separate commits.

Comment (onderkalaci) [reply src/backend/distributed/executor/executor.c]: Just removed those, I'll have another PR for indent etc.
